File: cvnode_base/cvnode_base.py
# Copyright 2022-2023 Antmicro <www.antmicro.com>
#
# SPDX-License-Identifier: Apache-2.0

"""Base class for computer vision nodes."""

from typing import List

from kenning_computer_vision_msgs.msg import SegmentationMsg
from kenning_computer_vision_msgs.srv import ManageCVNode, SegmentCVNodeSrv
from rclpy.node import Node
from rclpy.qos import QoSProfile
from sensor_msgs.msg import Image
from std_srvs.srv import Trigger


class BaseCVNode(Node):
    """
    Base class for tested computer vision nodes.

    Responsible for process with the CVNodeManager
    and running inference.
    """

    def __init__(self, node_name: str):
        """
        Initialize the node.

        Parameters
        ----------
        node_name : str
            Name of the node.
        """
        # Service client for node management
        self._manage_node_client = None

        # Service responsible for communication with the CVNodeManager
        self._prepare_service = None
        self._process_service = None
        self._cleanup_service = None

        super().__init__(node_name)

        self.registerNode('cvnode_register')

    def registerNode(self, manage_service_name: str):
        """
        Register node with the manage service.

        Parameters
        ----------
        manage_service_name : str
            Name of the service for nodes management.
        """
        self.get_logger().debug('Registering node')
        if self._manage_node_client:
            self._unregisterNode()

        self._manage_node_client = self.create_client(ManageCVNode,
                                                      manage_service_name)
        if not self._manage_node_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().error(
                'Node manage service not available')
            return

        prepare_service_name = f'{self.get_name()}/prepare'
        process_service_name = f'{self.get_name()}/process'
        cleanup_service_name = f'{self.get_name()}/cleanup'

        # Initialize services
        self._prepare_service = self.create_service(
            Trigger,
            prepare_service_name,
            self._prepareCallback
        )
        qos_profile = QoSProfile(depth=1)
        self._process_service = self.create_service(
            SegmentCVNodeSrv,
            process_service_name,
            self._processCallback,
            qos_profile=qos_profile)
        self._cleanup_service = self.create_service(
            Trigger,
            cleanup_service_name,
            self._cleanupCallback
        )

        # Create request for manage service
        request = ManageCVNode.Request()
        request.type = request.REGISTER
        request.node_name = self.get_name()
        request.prepare_srv_name = prepare_service_name
        request.process_srv_name = process_service_name
        request.cleanup_srv_name = cleanup_service_name

        def register_callback(future):
            response = future.result()
            self.get_logger().debug('Received register response')
            if not response:
                self.get_logger().error('Service call failed.')
                return

            if not response.status:
                error_msg = 'Register service call failed: ' + \
                    response.message
                self.get_logger().error(error_msg)
                return

            self.get_logger().debug('Register service call succeeded')
            return

        future = self._manage_node_client.call_async(request)
        future.add_done_callback(register_callback)

    def destroy_node(self):
        """
        Destroy node.

        Unregisters node if was registered.
        """
        self.get_logger().debug('Destroying node')
        if self._manage_node_client:
            self._unregisterNode()
        super().destroy_node()

    def prepare(self) -> bool:
        """
        Prepare node and model for inference.

        Returns
        -------
        bool :
            True if preparation was successful, False otherwise.
        """
        raise NotImplementedError

    def run_inference(self, X: List[Image]) -> List[SegmentationMsg]:
        """
        Run inference on the input data.

        Parameters
        ----------
        X : List[Image]
            List of input image messages.

        Returns
        -------
        List[SegmentationMsg] :
            List of postprocessed segmentation messages.
        """
        raise NotImplementedError

    def cleanup(self):
        """Cleanup allocated resources used by the node."""
        raise NotImplementedError

    def _unregisterNode(self):
        """Unregister node with service."""
        self.get_logger().debug('Unregistering node')
        if self._manage_node_client is None:
            self.get_logger().warn('Node was not registered')
            return
        request = ManageCVNode.Request()
        request.type = request.UNREGISTER
        request.node_name = self.get_name()

        self._manage_node_client.call_async(request)
        self._manage_node_client = None
        self.get_logger().debug('Node unregistered')

    def _prepareCallback(self, request: Trigger.Request,
                         response: Trigger.Response) -> Trigger.Response:
        """
        Callback for the prepare service.

        Responsible for preparing node for inference.

        Parameters
        ----------
        request : Trigger.Request
            Request for the prepare service.
        response : Trigger.Response
            Processed response for the prepare service client.

        Returns
        -------
        Trigger.Response :
            Processed response for the prepare service client.
        """
        self.get_logger().debug('Preparing node')
        if not self.prepare():
            self.get_logger().error('Node preparation failed')
            response.success = False
            return response
        response.success = True
        self.get_logger().debug('Node prepared')
        return response

    def _processCallback(self, request: SegmentCVNodeSrv.Request,
                         response: SegmentCVNodeSrv.Response
                         ) -> SegmentCVNodeSrv.Response:
        """
        Callback for the process service.

        Responsible for running inference on the input data.

        Parameters
        ----------
        request : SegmentCVNodeSrv.Request
            Request for the process service.
        response : SegmentCVNodeSrv.Response
            Processed response for the process service client.

        Returns
        -------
        SegmentCVNodeSrv.Response :
            Processed response for the process service client.
        """
        self.get_logger().debug('Executing inference on input data')
        response.output = self.run_inference(request.input)
        response.success = True
        self.get_logger().debug('Inference executed')
        return response

    def _cleanupCallback(self, request: Trigger.Request,
                         response: Trigger.Response) -> Trigger.Response:
        """
        Callback for the cleanup service.

        Responsible for cleaning up node's resources.

        Parameters
        ----------
        request : Trigger.Request
            Request for the cleanup service.
        response : Trigger.Response
            Processed response for the cleanup service client.

        Returns
        -------
        Trigger.Response :
            Processed response for the cleanup service client.
        """
        self.get_logger().debug('Cleaning up node')
        self.cleanup()
        response.success = True
        return response

File: src/cvnode_base.cpp
// Copyright 2022-2024 Antmicro <www.antmicro.com>
//
// SPDX-License-Identifier: Apache-2.0

#include "cvnode_base/cvnode_base.hpp"

namespace cvnode_base
{

using ManageCVNode = kenning_computer_vision_msgs::srv::ManageCVNode;
using SegmentCVNodeSrv = kenning_computer_vision_msgs::srv::SegmentCVNodeSrv;

BaseCVNode::BaseCVNode(const std::string &node_name, const rclcpp::NodeOptions &options) : Node(node_name, options)
{
    register_node("cvnode_register");
}

void BaseCVNode::prepare_callback(
    [[maybe_unused]] const std_srvs::srv::Trigger::Request::SharedPtr request,
    std_srvs::srv::Trigger::Response::SharedPtr response)
{
    RCLCPP_DEBUG(get_logger(), "Received request to prepare the node");
    if (!prepare())
    {
        RCLCPP_ERROR(get_logger(), "Error while preparing the node");
        response->success = false;
        cleanup();
        unregister_node();
        return;
    }
    response->success = true;
    RCLCPP_DEBUG(get_logger(), "Prepared the node");
    return;
}

void BaseCVNode::process_callback(
    const SegmentCVNodeSrv::Request::SharedPtr request,
    SegmentCVNodeSrv::Response::SharedPtr response)
{
    RCLCPP_DEBUG(get_logger(), "Received request to process input data");
    response->segmentation = run_inference(request->input);
    response->success = true;
    RCLCPP_DEBUG(get_logger(), "Processed the input data");
}

void BaseCVNode::cleanup_callback(
    [[maybe_unused]] const std_srvs::srv::Trigger::Request::SharedPtr request,
    std_srvs::srv::Trigger::Response::SharedPtr response)
{
    RCLCPP_DEBUG(get_logger(), "Cleaning up the node");
    cleanup();
    response->success = true;
    return;
}

void BaseCVNode::register_callback(const rclcpp::Client<ManageCVNode>::SharedFuture future)
{
    auto result = future.get();
    RCLCPP_DEBUG(get_logger(), "Received response from the register node service");
    if (!result)
    {
        RCLCPP_ERROR(get_logger(), "No response from the register node service");
        return;
    }
    if (!result->status)
    {
        RCLCPP_ERROR(get_logger(), "The node is not registered");
        RCLCPP_ERROR(get_logger(), "Error message: %s", result->message.c_str());
        cleanup();
        manage_client.reset();
        prepare_service.reset();
        process_service.reset();
        cleanup_service.reset();
        return;
    }

    RCLCPP_DEBUG(get_logger(), "The node was registered: %s", result->message.c_str());
}

void BaseCVNode::register_node(const std::string &manage_node_name)
{
    RCLCPP_DEBUG(get_logger(), "Registering the node");
    if (manage_client)
    {
        unregister_node();
    }

    manage_client = create_client<ManageCVNode>(manage_node_name);
    if (!manage_client->wait_for_service(std::chrono::seconds(1)))
    {
        RCLCPP_ERROR(get_logger(), "The node management service is not available after waiting");
        return;
    }

    auto request = std::make_shared<ManageCVNode::Request>();
    request->type = request->REGISTER;
    request->node_name = std::string(get_name());
    request->prepare_srv_name = std::string(get_name()) + "/prepare";
    request->process_srv_name = std::string(get_name()) + "/process";
    request->cleanup_srv_name = std::string(get_name()) + "/cleanup";

    prepare_service = create_service<std_srvs::srv::Trigger>(
        request->prepare_srv_name,
        std::bind(&BaseCVNode::prepare_callback, this, std::placeholders::_1, std::placeholders::_2));

    // Create QOS with keep last 1 to avoid queueing messages
    rmw_qos_profile_t qos = rmw_qos_profile_default;
    qos.depth = 1;

    process_service = create_service<SegmentCVNodeSrv>(
        request->process_srv_name,
        std::bind(&BaseCVNode::process_callback, this, std::placeholders::_1, std::placeholders::_2),
        qos);
    cleanup_service = create_service<std_srvs::srv::Trigger>(
        request->cleanup_srv_name,
        std::bind(&BaseCVNode::cleanup_callback, this, std::placeholders::_1, std::placeholders::_2));
    manage_client->async_send_request(
        request,
        [this](const rclcpp::Client<ManageCVNode>::SharedFuture future) { register_callback(future); });
}

void BaseCVNode::unregister_node()
{
    RCLCPP_DEBUG(get_logger(), "Unregistering the node");
    if (!manage_client)
    {
        RCLCPP_DEBUG(get_logger(), "The node is not registered");
        return;
    }

    auto request = std::make_shared<ManageCVNode::Request>();
    request->type = request->UNREGISTER;
    request->node_name = std::string(get_name());
    manage_client->async_send_request(request);
    manage_client.reset();
    prepare_service.reset();
    process_service.reset();
    cleanup_service.reset();
    RCLCPP_DEBUG(get_logger(), "The node was unregistered");
}

} // namespace cvnode_base
